### ISSUE TITLE
rogue public-key attack on BLS

### ISSUE THREAD

**Issue by Gilthoniel** [dedis/cothority#1675]
@kc1212 pointed out that the attack described here:

https://crypto.stanford.edu/~dabo/pubs/papers/BLSmultisig.html

could be an issue. We are safe for rosters generated from scratch but an attacker could join the cothority after learning about the current public keys. The attack is limited because one bad node can only forge one signature but still.

We need some help from the crypto guys to make sure we're safe.

**Comment by kc1212** [comment]
See the "The rogue public-key attack" section

**Comment by kc1212** [comment]
A paper reference in the link is https://eprint.iacr.org/2007/264.pdf
it mentioned proof of "knowledge of secret key" (KOSK)
what I understand it you can create a proof that shows you generated the public key correctly
the attacker won't be able to do this because he'll have to generate the public key in a special way

**Comment by jeffallen** [comment]
The current onet code does a challenge/response. The client does a normal TLS hello, sending a nonce. The server says, "hello, my public key is X and your nonce will be N, and you can tell I have the secret key x because I've signed your nonce". The client says, "I am public key Y and I've signed your nonce with secret key y."

The signature is generated with schnorr.Sign(), and it works even when the server's private key is a bn256.g2 key.

**Comment by kc1212** [comment]
yeah so in addition to checking the generated signature, we may need to check that the BLS public key is correctly generated i.e., `g^a` and not `g^b * victim_pk^-1`.

**Comment by jeffallen** [comment]
We've decided the defence against rogue public keys has to happen at the moment that a public key is accepted into the roster, i.e. at chain creation, or during a roster modification. And our analysis is that this will be possible to implement in a backwards compatible way after v3 is released.

**Comment by Gilthoniel** [comment]
Unfortunately it turns out we can't solve this without asking for a proof or by changing the BLS algorithm scheme which means the backwards compatibility of the fix is hard.

One way to prevent the attack is to use the improved scheme from the paper in the first comment that basically uses coefficient (public-key hashed to a number between 1 and 2^128) to break the forged keys. It's simple to implement but the current signatures wouldn't be verified anymore and if only a subset of the conodes are up-to-date, it doesn't work either.. I was thinking of using two signatures but it is not robust against downgrade attacks.

The other is to provide a proof by signing the public key with the private key (or anything else than the public key). The proof can be done once and kept forever because it's enough to prove that the private key is known. Again the main concern here is that old conodes won't have any of the new protocol that must be used to provide this proof. Actually proofs because you need to prove any BLS public key (and it should be two at the moment). This verification could be done only when the roster is gonna change and there we could think about adding a field to the skipblocks with the PoP (proof of possession) of any new member. That assumes the new member will be up-to-date as far as the leader is. It also means it needs to be online to be added. The leader would work to get this proof by the new conode and then for the forward link level 0 creation, the conodes can assert that a new conode has correct public keys.

**Comment by Gilthoniel** [comment]
Approach 1:
Because we're in a permission-full blockchain, there is a step where you add a conode in the cothority (i.e. using bcadmin). This could be used to exchange a proof of possession of the private key of every BN256 key pair. It would also be necessary to ask for PoPs when creating a genesis block.
+ backwards compatible
+ easy to implement
- dangerous if the roster change from another tool
- conodes must be online and responsive

**Comment by Gilthoniel** [comment]
Approach 2:
The modified BLS scheme uses coefficients to prevent a forged public key to be used. Key pairs could be used that way without any proof of possession.

[+] Don't require any proof of possession and protect any usage of a BLSCoSi
[+] Small overhead (well multiplications of a 128bits integer still), the aggregate of the public key can be cached because the coefficients are mapped from the public key using a hash function

[-] Backwards compatibility requires some tricks to prevent a downgrade attack
[-] Need to implement a kyber.Scalar over 128 bits (Go, JS, Java)
[-] Need to update all libraries
[-] V3 might need double signature verification (until V4 where we can remove the old scheme)

**Comment by Gilthoniel** [comment]
Approach 3:
Append PoP during the BLSCoSi protocol. This might sounds overkill *but* one benefit of a PoP is that you can reuse it indefinitely because you only want to know if the private key exists meaning that you can store it. So the announcement message could include a mask of required PoP so that eventually this mask would be empty and the PoP is transmitted only once to every node.

[+] Any key pair added to the roster in any way will be verified
[+] Very small overhead (PoP sent once for each conode)
[+] No further crypto implementation required

[-] Backwards compatibility requires some tricks to prevent a downgrade attack (basically either you ignore a missing PoP or you force it by a parameter passed to the conode when you know your conodes are up-to-date)
[-] Might need to think how to store the PoPs to manage the size

**Comment by Gilthoniel** [comment]
I would go for approach 3 because it's more robust than approach 1 and it looks like it could be done really small changes. Approach 2 seems really complex compared to the other two so we might keep it for V4..

**Comment by jeffallen** [comment]
I agree that approach 3 looks promising, and perhaps even deserves to be published more widely.

**Comment by Gilthoniel** [comment]
So after thinking this through, it's not enough to protect the signing protocol because it doesn't prevent a rogue to enter the roster, only to use it to collectively sign but as soon as the rogue key is inside, nothing prevent from propagating a forged block.

Therefore I guess the most important security here is to prevent any rogue public-key to enter the roster at the skipchain service level so that if a roster is registered in a block you know for sure it has been verified.

I'll keep the idea of Approach 3 to propagate the Proof of Possession to reduce the load of the network but the verification will be extended to verify new conodes only (i.e. each service with a BN256 key pair)

**Comment by ineiti** [comment]
Couldn't you do approach 1 in the `Config` contract? The contract would allow new nodes only if there is a PoP (Proof of Personhood?). For re-ordering the roster, there would be no need to have this proof.

Or simply add an optional field to `ServiceIdentity`, and then the config contract would only allow new nodes if that field is present and correct.

**Comment by Gilthoniel** [comment]
PoP == Proof of Possession in that context. I feel like it's more the role of the skipchain service to make this verification. Verifying the roster is part of forward link level 0 signature procedure.

**Comment by ineiti** [comment]
When we wrote the skipchain service, the goal was to have `Verification` functions that can check different things for each new block. Specifically checking whether the `Roster` is valid with regard to this attack looks like a good use-case for these `Verification` functions, as theoretically the skipchain service could also run with another consensus-algorithm.

As `ByzCoin` is in fact one of the `Verification` functions, it would make sense to verify in the `ByzCoinVerification` function whether the roster is valid. And this function will call the config-contract if the roster is updated. So from that point of view it looks OK to put it in the config contract.

Or if you decide to add an optional field to the `ServiceIdentity`, you can create a `BLSVerification` method that checks the roster for a new block in the skipchain and refuses if the public keys don't match. But then you'll have troubles because the public key is in the `ServiceIdentity["byzcoin"]`, and not in the `ServiceIdentity["skipchain"]`. So this should move then.

TLDR: for me it makes sense to put it in the configuration-contract ;)

**Comment by Gilthoniel** [comment]
Whichever consensus is used is not really the question here because the purpose is to prevent any forged public key to enter the roster, not only the one associated with the skipchain service. That's why this is so important that it should not be part of an optional verification function (I know, the base one is always included ;) ).

Also for the ServiceIdentity, I'd like to avoid new fields as the Roster moves around in the network a lot and then the smaller the better.

**Comment by ineiti** [comment]
Well if you want to have a `one solution fits all`, I don't see how you can avoid adding a field to the `ServiceIdentity`. What I understand from your proposition is very service-specific and would have to be repeated for every service that wants to use BLS signatures. Or did I get something wrong here?

**Comment by Gilthoniel** [comment]
I was planning to use the immutability property of the blockchain so that if a roster has been approved in a block, it can be assumed later on as verified and not maliciously forged.

**Comment by ineiti** [comment]
OK, so it's not a `one solution fits all`, in the sense that a `BLSCoSi` service or protocol does not look the roster up in the blockchain, so it cannot use it.

But if you use the `ServiceIdentity` approach, you can use it in any service where you want to make sure that the public keys are correct.

**Comment by Gilthoniel** [comment]
You can't protect the protocol. For example:
1. Someone wants to use the BLSCoSi service to sign a file using a roster found somewhere
2. It happens that one of the member has a forged public-key for the entire roster
3. It happens that it is the chosen server where the request is sent
4. The server ignores the protocol and simply forge a signature
5. Profit

What I mean is that the only way other than approach 2 is to make sure the roster is correct, either by using the blockchain or by manually verifying it. I'm making a PKI service that can be used in that way which is better than the `ServerIdentity` approach I think because the proof can be stored so you don't need to pollute the roster structure.

**Comment by ineiti** [comment]
You don't even need to run a protocol, you can setup your own roster with yourself as the leader, and then sign without sending any messages at all...

Please have a look at #1797 so we can compare it to #1794

**Comment by jeffallen** [comment]
While working through all the arguments on all the sides of all of the different approaches, I've come to the conclusion that fixing this in an interactive way due to backwards compatibility is messy, and will leave mess we have to support behind it. The best solution we've identified is https://github.com/dedis/cothority/issues/1675#issuecomment-478503423, which is to exclude the rouge keys non-interactively, instead of moving around proofs of knowledge, and storing them, etc.

Deploying a new signature protocol (for discussion, let's call it BlsCosiWithCoeff) as the default for new skipchains made starting with v3.1, and supporting skipchains running on BlsCosi for existing skipchains means that new skipchains will not be at risk from rogue public keys. The only production skipchain which exists (the DEDIS cothority) will remain at risk, but we know that all the public keys in it were honestly computed, and we could even get and publish proof of knowledge of secret keys for them as well, and require it (as an out of band process) while evolving the roster of that chain.

We also know that we need to sooner or later consider how to teach skipchains to evolve their co-signing protocol. Solving this issue by deploying a new BlsCosiWithCoeff signing protocol in parallel with BlsCosi will start us on the road to having co-signing agility in an existing Skipchain.
